[bugfix] Improve sync states management for checkpointing (#334)
PyTorchLightning/metrics: docs/source/references/modules.rst

.. autoclass:: torchmetrics.Metric
    :noindex:

.. note::

>     Metric states are being synchornized by default in a distributed setting when using :meth:`~torchmetrics.Metric.state_dict`.

Reviewer: Maybe it is worth mentioning that the state dict by default is empty because `persistent=False`

Author: Done !

    This behaviour can be blocked in the following ways.
    We strongly recommend calling :meth:`~torchmetrics.Metric.state_dict` on all processes or this can cause a deadlock.

.. doctest::

    >>> from torchmetrics import Metric
    >>> metric = Metric(should_sync_state_dict=False)
    >>> metric.state_dict()

> .. doctest::

Reviewer 2: maybe explain the difference here or when to use which way?

Author: Done !

.. autoclass:: torchmetrics.Metric
    :noindex:

> .. note::
> 
>     Metric states are being synchornized by default when ``persistent=True`` in a distributed setting
>     when using :meth:`~torchmetrics.Metric.state_dict`
>     This behaviour can be blocked in the following ways.
>     We strongly recommend calling :meth:`~torchmetrics.Metric.state_dict` on all processes or this can cause a deadlock.

Reviewer 3: I don't think this should be a note, it's an important detail

```suggestion
Metric states are synchornized when ``persistent=True`` in a distributed setting
when calling :meth:`~torchmetrics.Metric.state_dict`
We strongly recommend calling :meth:`~torchmetrics.Metric.state_dict` on all processes, otherwise this can cause a deadlock.
This can also happen when the ``state_dict`` of your :class:`torch.nn.Module` is generated if metric attributes have been registered.
```

> We recommend to use ``dist_sync_state_dict`` when your metric is nested inside another :class:`~torch.nn.Module`.

Reviewer 3: ```suggestion
The synchronization can be avoided in the following ways.

Set ``dist_sync_state_dict`` when your metric is nested inside another :class:`torch.nn.Module`.
```

Author: Done !

.. doctest::

    >>> import torch
    >>> from torchmetrics import Metric
    >>> class AverageMeter(Metric):
    ...     def __init__(self):
    ...         super().__init__(dist_sync_state_dict=False)
    ...         self.add_state("x", torch.tensor(0.0), dist_reduce_fx=torch.sum, persistent=True)
    ...         self.add_state("total", torch.tensor(0.0), dist_reduce_fx=torch.sum, persistent=True)
    ...
    ...     def update(self, x: torch.Tensor):
    ...         self.x += x
    ...         self.total += 1
    ...
    ...     def compute(self):

>     ...         return self.x / self.total

Reviewer 3: Some whitespace

```suggestion
    ...         return self.x / self.total
    ...
```

    >>> metric = AverageMeter()
    >>> metric.state_dict()
    OrderedDict(...)

> We recommend to use ``state_dict(should_sync=False)`` if you can easily access your :class:`~torchmetrics.Metric`.

Reviewer 3: ```suggestion
Or use ``state_dict(should_sync=False)`` if you can easily access your :class:`~torchmetrics.Metric`.
```
